doc/generated_testcases_v2.yaml

```diff
@@ -0,0 +1,221 @@
+# Example generators.yaml
+
+# The solution is used to generate a .ans for each generated .in which doesn't
+# yet have a corresponding .ans. If there are generators that don't write a .ans
+# file themselves, this must be specified.
+# This should read the input from stdin and write to stdout.
+#
+# Absolute paths are taken relative to the problem root, relative paths are
+# taken relative to the generators directory.
+#
+# TOOLING: may pick a default if not specified, but should raise a warning.
+solution: /submissions/accepted/sol.py
+
+# The visualizer is used when no suitable image was generated already.
+# This should read {name}.in and/or {name}.ans from the current working
+# directory, and write {name}.ext for an extension in:
+# .png, .jpg, jpeg, .svg
+#
+# TOOLING: may provide a flag to make running this optional, as it can be slow
+# and usually isn't required.
+visualizer: /visualizers/asy.py {name}
+
+# Optionally, a salt for generating the {seed} variables. Will be prepended to
+# the command being run.
+random_salt: abcd
+
+# We support three types of generators:
+# - Standalone files, like generators/a.cpp, generators/b.py, ..., which will
+#   be compiled and run the same way as submissions.
+# - Directories, like 
+#   - generators/gen/tree.cpp
+#   - generators/gen/lib.h
+#   This will be compiled and run the same was as directory validators. Build
```

Comment:
The specification for validators **requires** a `run` script/executable, which *may* be generated by an optional `build` script/executable. I think for consistency it would be good to stick to this.
This is anyways necessary to prevent ambiguity of how to build/run this.

Reply:
Hmm I don't think that's true, see [Programs](https://clics.ecs.baylor.edu/index.php?title=Problem_format#Programs). 
The table there also lists the main entrypoint to prevent ambiguity.

> Language is determined by looking at the file endings

is not very specific, but problemtools picks the language matching the most files, and breaks ties on priority, as [specified](https://github.com/Kattis/problemtools/blob/develop/problemtools/config/languages.yaml#L27-L42) in `languages.yaml`. I implemented the same for BAPCtools.

The entrypoint will be the single file, or a file matching `main`, or the lexicographically smallest file, see [here](https://github.com/Kattis/problemtools/blob/develop/problemtools/config/languages.yaml#L79-L86).

Thinking about it, that still doesn't really work for manually specified generators:
```
generators:
   trees:
      - trees.py
      - lib.py
```
would get `lib.py` as the entrypoint because `'lib.py' < 'trees.py'`. 
**What about making the first item in this list the entry point? I think that's quite natural anyway.**

Side note: this part of the spec already mentions 'POSIX-complient scripts' so we can probably refer to POSIX-shell parsing as well.

Reply:
Hmmm... in DOMjudge we use these stricter rules about build/run scripts, and I thought these were copied into this spec. I really think this leaves too much ambiguity: what if the code in a directory might compile to two binaries? And I don't think that lexicographically first is natural enough to use as an implicit rule. If it was easy to have a single binary/entry point, then you could have used a single-file generator/validator in the first place.

For declaring the first element in the YAML list the entry point: that only works if the entry point is a file, but not for example if it is a Java class, of which you could have multiple per file. But since this method of specifying multi-file generators is really a convenience-method, I'd be willing to accept some ambiguity here by leaving the details up to the tooling.

Reply:
Well, we've used DOMjudge in the past without `run` files:
```
bapc/jazzitup % tree output_validators 
output_validators
└── output_validator
    ├── output_validator.cpp
    └── validation.h
```

If you have a `build` file, you should probably have/build a `run` file as well, but having neither works fine as long as it's a single binary and the language is clear.

I think the usual case is still to have a single binary/executable/entry point with some library dependencies, where a directory doesn't work but auto-language-detection is still fine.

Current language detection puts the `{mainclass}` as the [filename without extension](https://github.com/Kattis/problemtools/blob/develop/problemtools/config/languages.yaml#L88-L90), and that should probably work in most cases.